src/rust.rs

```diff
@@ -463,6 +465,196 @@ macro_rules! rooted {
+    }
+
+    /// Used by `CustomAutoTraceable` implementer to trace its contents.
+    /// Corresponds to virtual `trace` call in a `CustomAutoRooter` subclass (C++).
+    fn do_trace(&self, trc: *mut JSTracer);
+}
+
+unsafe impl<T: CustomTrace> CustomAutoTraceable for CustomAutoRooter<T> {
+    fn do_trace(&self, trc: *mut JSTracer) {
+        self.data.trace(trc);
+    }
+}
+
+impl<T: CustomTrace> CustomAutoRooter<T> {
+    pub fn new(data: T) -> Self {
+        CustomAutoRooter {
+            _base: jsapi::CustomAutoRooter {
+                _vftable: &<Self as CustomAutoTraceable>::vftable,
+                _base: AutoGCRooter::new_unrooted(AutoGCRooterTag::CUSTOM),
+            },
+            data,
+        }
+    }
+
+    pub fn root<'a>(&'a mut self, cx: *mut JSContext) -> CustomAutoRooterGuard<'a, T> {
+        CustomAutoRooterGuard::new(cx, self)
+    }
+}
+
+/// An RAII guard used to root underlying data in `CustomAutoRooter` until the
+/// guard is dropped (falls out of scope).
+/// The underlying data can be accessed through this guard via its Deref and
+/// DerefMut implementations.
+/// This structure is created by `root` method on `CustomAutoRooter` or
+/// by an appropriate macro (e.g. `rooted_vec!` for `SequenceRooter` alias).
+pub struct CustomAutoRooterGuard<'a, T: 'a + CustomTrace> {
+    rooter: &'a mut CustomAutoRooter<T>
+}
+
+impl<'a, T: 'a + CustomTrace> CustomAutoRooterGuard<'a, T> {
+    pub fn new(cx: *mut JSContext, rooter: &'a mut CustomAutoRooter<T>) -> Self {
+        unsafe {
+            rooter.add_to_root_stack(cx);
+        }
+        CustomAutoRooterGuard {
+            rooter
+        }
+    }
+
+    pub fn handle(&self) -> Handle<T> where T: RootKind {
+        unsafe {
+            Handle::from_marked_location(&self.rooter.data)
+        }
+    }
+
+    pub fn handle_mut(&mut self) -> MutableHandle<T> where T: RootKind {
+        unsafe {
+            MutableHandle::from_marked_location(&mut self.rooter.data)
+        }
+    }
+}
+
+impl<'a, T: 'a + CustomTrace> Deref for CustomAutoRooterGuard<'a, T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        &self.rooter.data
+    }
+}
+
+impl<'a, T: 'a + CustomTrace> DerefMut for CustomAutoRooterGuard<'a, T> {
+    fn deref_mut(&mut self) -> &mut T {
+        &mut self.rooter.data
+    }
+}
+
+impl<'a, T: 'a + CustomTrace> Drop for CustomAutoRooterGuard<'a, T> {
+    fn drop(&mut self) {
+        unsafe {
+            self.rooter.remove_from_root_stack();
+        }
+    }
+}
+
+pub type SequenceRooter<T> = CustomAutoRooter<Vec<T>>;
+pub type SequenceRooterGuard<'a, T> = CustomAutoRooterGuard<'a, Vec<T>>;
+
+#[macro_export]
+macro_rules! rooted_seq {
+    (in($cx:expr) let $name:ident = $init:expr) => {
+        let mut __root = $crate::rust::CustomAutoRooter::new($init);
+        let $name = __root.root($cx);
+    };
+    (in($cx:expr) let mut $name:ident = $init:expr) => {
+        let mut __root = $crate::rust::SequenceRooter::new($init);
```

Comment:
One of these is CustomAutoRooter, and one is SequenceRooter. That's surprising!

Reply:
Oops! Fixed that now, renamed the macro to a more general `auto_root` name, to be used with CustomAutoRooter itself.